### ISSUE TITLE
Deprecate and remove synthetic `package:flutter_gen`

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#102983]
Update as of 2024-02-18: https://docs.flutter.dev/release/breaking-changes/flutter-generate-i10n-source.

---

Currently, `flutter_gen` is a virtual package we generate to provide certain functionality (e.g. [internationalisation](https://docs.flutter.dev/development/accessibility-and-localization/internationalization#adding-your-own-localized-messages)). Developers have to import that package into their app. The package itself is not listed in their app's `pubspec.yaml`, though. Instead, the `flutter_gen` package is made available by a rewrite of the package config file itself.

This "hack" has caused issues with multiple tools in the past, e.g.:

* build_runner (https://github.com/dart-lang/build/issues/2835)
* linter (specifically, the `depend_on_referenced_packages ` lint, https://github.com/dart-lang/sdk/issues/58690)
* Dart Code (https://github.com/Dart-Code/Dart-Code/issues/3948)
* pub (mentioned in https://github.com/dart-lang/sdk/issues/58690)

It's expensive for tooling to have to handle package:flutter_gen as a special case and it's easy to overlook that special support for it is needed resulting in a sub-par user experience.

We should rethink whether there are better ways to support the functionality that `flutter_gen` is currently providing that's more aligned with the dart ecosystem and its tooling. Maybe the upcoming support for [macros](https://github.com/dart-lang/language/blob/master/working/macros/feature-specification.md) in Dart can help with that).

**Comment by goderbauer** [comment]
Open question:
* what do we use flutter_gen for? Is it just internationalisation?

**Comment by jakemac53** [comment]
build_runner also has experienced issues with this package, https://github.com/dart-lang/build/issues/2835 probably has the most context, but basically we need more information about the dependency graph that is provided by the package config, so we require valid pubspec files.

https://github.com/dart-lang/build/issues?q=is%3Aissue+flutter_gen+is%3Aclosed is a link to all the issues that have come up, people seem to still be struggling with this on a regular basis and chime in on the closed issues.

**Comment by jonahwilliams** [comment]
rethink = remove right?

Step 1: warn users that have it they need to opt out by VERSION like a standard breaking change
Step 2: make the default to generate to source
Step 3: remove the ability to use flutter_gen

Dates TBD

**Comment by zanderso** [comment]
Build systems like the tool's build system typically become very confused when the source tree is modified during dependency/modification calculations and/or during building. This can be sidestepped by generating code into the source tree during a "repo setup" step like `flutter packages get` that is separate from the build. This has the drawback that it is not compatible with regenerating source during a hot reload or a hot restart. For example, if you modify the i18n data file, and expect the build system to notice that it's dirty, and that source needs to be regenerated in response, then that regenerated source code can't go in the source tree. Rather it has to go into a separate build output tree.

Ideally, there would be a general way to teach Dart tooling about source code generated outside of the source tree since that is a general property of build systems and not really specific to the Flutter build.

tl;dr: the tool can generate source into the source tree, but hot reload/restart won't be supported.

**Comment by gbassisp** [comment]
> rethink = remove right?
> 
> Step 1: warn users that have it they need to opt out by VERSION like a standard breaking change Step 2: make the default to generate to source Step 3: remove the ability to use flutter_gen
> 
> Dates TBD

What about gen-l10n? Would it be removed too? Then what's the internationalization tool we have to generate from ARB?

**Comment by AlexV525** [comment]
As for today, the package is lacking maintenance which might block users from upgrading (version conflict on `analyzer`).

**Comment by christopherfujino** [comment]
> As for today, the package is lacking maintenance which might block users from upgrading (version conflict on `analyzer`).

Can you give more details? Is there an independent tracking issue for this?

**Comment by AlexV525** [comment]
> Can you give more details? Is there an independent tracking issue for this?

https://github.com/FlutterGen/flutter_gen/issues/414

**Comment by christopherfujino** [comment]
> > Can you give more details? Is there an independent tracking issue for this?
> 
> [FlutterGen/flutter_gen#414](https://github.com/FlutterGen/flutter_gen/issues/414)

Ahh, that is a different package than what this issue is referring to. This issue is tracking a synthetic package that the flutter tool auto-generates when the user uses localizations to inject the generated code into their flutter app. See https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/localizations/gen_l10n.dart#L21

**Comment by AlexV525** [comment]
> Ahh, that is a different package than what this issue is referring to. This issue is tracking a synthetic package that the flutter tool auto-generates when the user uses localizations to inject the generated code into their flutter app. See https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/localizations/gen_l10n.dart#L21

Thanks for clarifying. I'll mark the above comments as off-topic then.

**Comment by astronomico** [comment]
[For my case - simple solution](https://www.appsloveworld.com/flutter/200/59/build-runner-package-in-my-flutter-app-for-json-serialization-not-working)
I only have to remove flutter generate specification on pubspec.yaml with my localizations builds

#The following section is specific to Flutter packages.
flutter:
    # generate: true // remove this

**Comment by matanlurey** [comment]
We've agreed to work on this, going with the "generate into source" strategy, as part of https://github.com/flutter/flutter/issues/157819.

**Comment by pcj** [comment]
Thank you for working on this!  As a regular user, this was a significant pain point.

**Comment by matanlurey** [comment]
There is still work to do, the flag that turns it off by default has not been enabled yet (it's blocked on unrelated work that is WIP: https://github.com/flutter/flutter/issues/160257.

**Comment by matanlurey** [comment]
There is not a migration planned, but there is an announcement including what to do (if anything):

https://docs.flutter.dev/release/breaking-changes/flutter-generate-i10n-source

**Comment by BuyMyBeard** [comment]
> There is not a migration planned, but there is an announcement including what to do (if anything):

How are generated files generated now? After doing the migration, I can't seem to get the new files to generate.
Do I also need to change all my flutter_gen import paths?

**Comment by matanlurey** [comment]
As of the latest master (https://github.com/flutter/flutter/pull/160289), this is true by default.

After the next stable release, the ability to opt-out will be removed, and the feature can be removed.

**Comment by giorgio79** [comment]
Please clarify in the migration steps if we now need to remove these import lines?

`import 'package:flutter_gen/gen_l10n/app_localizations.dart'; 
`
I am getting tons of errors due to
error: Undefined name 'AppLocalizations'. (undefined_identifier at ....)

I am trying method 1, whereby we add 

`synthetic-package: false
`
in l10.yaml

**Comment by BuyMyBeard** [comment]
> Please clarify in the migration steps at https://docs.flutter.dev/release/breaking-changes/flutter-generate-i10n-source if we now need to remove these import lines?


Basically, all this does is change the package where the localizations are generated. by default, you should find them in `lib/i18n`. You have to replace the import path by your new path.

**Comment by Gustl22** [comment]
[Option 1](https://docs.flutter.dev/release/breaking-changes/flutter-generate-i10n-source#migration-guide) breaks `dart format . --set-exit-if-changed` checks (as commonly used in pipelines), as the generated file does not the `dart format` rules. We should add a `// dart format off` in the generated files. As currently dart cannot exclude specific files from being formatted.
_It's independent from if the files are checked into git or not._
Simply formatting the generated file before finishing wouldn't solve this also, as users might have custom project formatting with `dart format . -l 120` option.

See: https://github.com/Dart-Code/Dart-Code/issues/4482#issuecomment-2707399548
And: https://github.com/Dart-Code/Dart-Code/issues/4482#issuecomment-2707399548

**Comment by esodot** [comment]
I guess we will need to have an update in the [flutter i18n documentation]("https://docs.flutter.dev/ui/accessibility-and-internationalization/internationalization"), mentioning that the setting "synthetic-package: false" in l10n.yaml has to be in place as well for the time being.

**Comment by Gustl22** [comment]
As Ref for updating the docs: https://github.com/flutter/website/issues/11765

**Comment by Gustl22** [comment]
~@matanlurey despite being closed, the docs still need adapting:~
https://github.com/flutter/website/issues/11765

Edit: Seems no more to be an issue with flutter 3.32.x => https://github.com/flutter/website/pull/12054/files

**Comment by Gustl22** [comment]
@matanlurey sorry for annoying you again. This issue is still referenced for removal of code in current code base:
https://github.com/search?q=repo%3Aflutter%2Fflutter%20102983&type=code

_Everything kinda got mixed as some things didn't work with 3.28.x but did with synthetic package (and wasn't documented before explictly), and now it does work with 3.32.x (especially with problems workspaces, enabling SwiftPM), e.g. see: https://github.com/flutter/flutter/issues/169209#issuecomment-2898601109. Sorry for bothering. Still hope this leads to improvements for the community_.

Thank you for maintaining / contributing in any case.
